pkg/resource/aggregator.go

```diff
@@ -0,0 +1,215 @@
+}
+
+func (a Aggregator) aggregateKibanaMemory() (resource.Quantity, error) {
+	var kbList kbv1.KibanaList
+	err := a.client.List(&kbList)
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+
+	var total resource.Quantity
+	for _, kb := range kbList.Items {
+		var mem resource.Quantity
+
+		// read the container memory limits
+		for _, container := range kb.Spec.PodTemplate.Spec.Containers {
+			if container.Name == kbv1.KibanaContainerName {
+				mem = *container.Resources.Limits.Memory()
+
+				// if not, fallback to the max JVM heap size
+				if mem.IsZero() {
+					for _, envVar := range container.Env {
+						if envVar.Name == kbconfig.EnvNodeOpts {
+							mem, err = memFromNodeOptions(envVar.Value)
+							if err != nil {
+								return resource.Quantity{}, err
+							}
+						}
+					}
+				}
+			}
+		}
+
+		// if not, fallback to the default limits
+		if mem.IsZero() {
+			mem = kbpod.DefaultMemoryLimits
+		}
+
+		total.Add(multiply(mem, kb.Spec.Count))
+		log.V(1).Info("Collecting", "kibana_name", kb.Name, "memory", mem.String(), "count", kb.Spec.Count)
+	}
+
+	return total, nil
+}
+
+func (a Aggregator) aggregateApmServerMemory() (resource.Quantity, error) {
+	var asList asv1.ApmServerList
+	err := a.client.List(&asList)
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+
+	var total resource.Quantity
+	for _, as := range asList.Items {
+		var mem resource.Quantity
+
+		// read the container memory limits
+		for _, container := range as.Spec.PodTemplate.Spec.Containers {
+			if container.Name == asv1.ApmServerContainerName {
+				mem = *container.Resources.Limits.Memory()
+			}
+		}
+
+		// if not, fallback to the default limits
+		if mem.IsZero() {
+			mem = apmserver.DefaultMemoryLimits
+		}
+
+		total.Add(multiply(mem, as.Spec.Count))
+		log.V(1).Info("Collecting", "as_name", as.Name, "memory", mem.String(), "count", as.Spec.Count)
+	}
+
+	return total, nil
+}
+
+// maxHeapSizePattern is the pattern to extract the max Java heap size (-Xmx<size>[g|G|m|M|k|K])
+const maxHeapSizePattern = "-Xmx([0-9]*)([gGmMkK]*)"
+
+var maxHeapSizeRe = regexp.MustCompile(maxHeapSizePattern)
+
+// memFromJavaOpts extracts the maximum Java heap size from a Java options string, multiplies the value by 2
+// and converts it to a resource.Quantity
+func memFromJavaOpts(javaOpts string) (resource.Quantity, error) {
+	match := maxHeapSizeRe.FindStringSubmatch(javaOpts)
+	if len(match) != 3 {
+		return resource.Quantity{}, fmt.Errorf("cannot extract max jvm heap size from %s", javaOpts)
+	}
+	value, err := strconv.Atoi(match[1])
+	if err != nil {
+		return resource.Quantity{}, err
+	}
+	// capitalize the suffix unless it's a K to have a surjection of [g|G|m|M|k|K] in [G|M|k]
+	suffix := match[2]
+	switch suffix {
+	case "K", "k":
+		suffix = "k"
+	default:
+		suffix = strings.ToUpper(suffix)
+	}
+	// multiply by 2 and convert it in a quantity using the suffix
+	return resource.ParseQuantity(fmt.Sprintf("%d%s", value*2, suffix))
```

Comment:
Aren't xmx/xms using binary prefixes? I.e.: `-Xmx1g` would translate to 1GiB (1,073,741,824 bytes) and here we will treat it as 1GB (1,000,000,000 bytes)?

Reply:
-Xmx1g would translate to 1GB (0.93 GiB).

`ParseQuantity` detects the format to use from the suffix:
```
//  Ki | Mi | Gi | Ti | Pi | Ei         => binarySI
//  m | "" | k | M | G | T | P | E      => decimalSI
```

Reply:
>

Reply:
I misread the question. You are right. -Xmx1g must be translated into 1GiB.